File: plugin/src/main/java/ru/arc/xserver/PlayerListAnnouncer.java
package ru.arc.xserver;

import com.google.gson.Gson;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import ru.arc.config.Config;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;


@RequiredArgsConstructor
public class PlayerListAnnouncer{



    final Config config;
    final RedisManager redisManager;
    final String channel;

    Gson gson = new Gson();
    Map<UUID, PlayerData> map = new ConcurrentHashMap<>();


    public void addPlayer(UUID uuid, String username, String server){
        map.put(uuid, new PlayerData(username, server, uuid, System.currentTimeMillis()));
    }

    public void updatePlayer(UUID uuid, String username, String server){
        PlayerData data = map.get(uuid);
        if(data != null){
            data.setServer(server);
        } else{
            addPlayer(uuid, username, server);
        }
    }

    public void removePlayer(UUID uuid){
        map.remove(uuid);
    }

    public void removeAllPlayers(){
        map.clear();
    }

    public void announce(){
        CompletableFuture.supplyAsync(() -> gson.toJson(map.values()))
                        .thenAccept(json -> redisManager.publish(channel, json));
    }


    @Data
    @AllArgsConstructor
    static class PlayerData{
        String username, server;
        UUID uuid;
        long joinTime;
    }

}

File: src/main/kotlin/ru/arc/xserver/PlayerListAnnouncer.kt
package ru.arc.xserver

import com.google.gson.Gson
import ru.arc.redis.RedisOperations
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class PlayerListAnnouncer(
    private val redisManager: RedisOperations,
    private val channel: String,
) {
    private val gson = Gson()
    private val map: MutableMap<UUID, PlayerData> = ConcurrentHashMap()

    fun addPlayer(uuid: UUID, username: String, server: String) {
        map[uuid] = PlayerData(username, server, uuid, System.currentTimeMillis())
    }

    fun updatePlayer(uuid: UUID, username: String, server: String) {
        val data = map[uuid]
        if (data != null) {
            data.server = server
            data.username = username
        } else {
            addPlayer(uuid, username, server)
        }
    }

    fun removePlayer(uuid: UUID) {
        map.remove(uuid)
    }

    fun removeAllPlayers() {
        map.clear()
    }

    fun announce() {
        val snapshot = map.values.map(PlayerData::copy)
        redisManager.publish(channel, gson.toJson(snapshot))
    }

    fun serverForUsername(username: String): String? =
        map.values.firstOrNull { it.username.equals(username, ignoreCase = true) }?.server

    data class PlayerData(
        var username: String,
        var server: String,
        val uuid: UUID,
        val joinTime: Long,
    )
}
